File: lambda_calc.py
#use this to test, for example N5(P1)(0) should be 5
P1 = lambda x: x + 1

# some set of combinators
I = lambda x: x
M = lambda x: x(x)
K = lambda a: lambda b: a
KI = lambda a: lambda b: b
C = lambda f: lambda a: lambda b: f(b)(a)
B = lambda f: lambda g: lambda a: f(g(a))
Th = lambda a: lambda f: f(a)
V = lambda a: lambda b: lambda f: f(a)(b)
B1 = lambda f: lambda g: lambda a: lambda b: f(g(a)(b))

#booleans True and False
T = lambda a: lambda b: a
F = lambda a: lambda b: b

#boolean functions, supply with T and F return T and F
Not = lambda p: p(F)(T)
And = lambda p: lambda q: p(q)(p)
Or = lambda p: lambda q: p(p)(q)
Beq = lambda p: lambda q: p(q)(Not(q))

#some hard coded church numerals
N0 = lambda f: lambda a: a 
N1 = lambda f: lambda a: f(a)
N2 = lambda f: lambda a: f(f(a))
N3 = lambda f: lambda a: f(f(f(a)))
N4 = lambda f: lambda a: f(f(f(f(a))))
N5 = lambda f: lambda a: f(f(f(f(f(a)))))
churchToInt = lambda n: n(P1)(0)

#lists and list operators
pair = lambda a: lambda b: lambda f: f(a)(b)
first = lambda p: p(K)
second = lambda p: p(KI)
phi = lambda p: pair(second(p))(succ(second(p)))
setFirst = lambda c: lambda p: pair(c)(second(p))
setSecond = lambda c: lambda p: pair(first(p))(c)

#chuch bool operations
succ = lambda n: lambda f: lambda a: f(n(f)(a))
add = lambda n: lambda k: lambda f: B(n(f))(k(f))
exp = lambda n: lambda k: k(n)
mult = lambda n: lambda k: lambda f: n(k(f))
pred = lambda n: first( n (phi) (pair(N0)(N0)))
sub = lambda n: lambda k: k(pred)(n)

#these functions take Church numerals and return church bools
isZero = lambda n: n(K(F))(T)
LeQ = lambda n: lambda k: isZero(sub(n)(k))
eQ = lambda n: lambda k: And(LeQ(n)(k))(LeQ(n)(k))

#lists
nil = pair(T)(T)
const = lambda h: lambda t: pair(F)(pair(h)(t))
head = lambda z: first(second(z)) 
tail = lambda z: second(second(z))

#binary trees
Empty = lambda e: lambda b: e
Branch = lambda x: lambda l: lambda r: lambda e: lambda b: b(x)(l(e)(b))(r(e)(b))
Treesum = lambda t: t(0)(lambda x:lambda l:lambda r: x+l+r)
#todo DFS BFS

File: lambda_calc.js
//use this function for testing, for example N5(P1)(0) should be 5
const P1 = x => x + 1

//combinators
const I = a => a
const M = a => a(a)
const K = a => b => a
const C = f => a => b => f(b)(a)
const B = f => g => a => f(g(a))
const Th = a => f => f(a)
const V = a => b => f => f(a)(b)
const B1 = f => g => a => b => f(g(a)(b))
const KI = a => b => b
const Y = f => M( x => f(M(x)))
const Z = f => M(x => f( v=> M(x)(v)))

//pair constructor and pair functions
const pair = a => b => f => f(a)(b)
const first = p => p(K)
const second = p => p(KI)
const phi = p => pair(second(p))(succ(second(p)))
const setFirst = c => p => pair(c)(second(p))
const setSecond = c => p => pair(first(p))(c)

//church booleans
const T = a => b => a
const F = a => b => b

//boolean functions, supply with T and F return T or F
const Not =  p => p(F)(T)
const And = p => q => p(q)(p)
const Or = p => q => p(p)(q)
const Beq = p => q => p(q)((Not(q)))

//church number operators
const succ = n => f => a => f(n(f)(a))
const add = n => k => f => B(n(f))(k(f))
const exp = n => k => k(n)
const pred = n => first(n(phi)(pair(N0)(N0)))
const sub = n => k => k(pred)(n)

//hard coded church numbers
const N0 = f => a => a
const N1 = f => a => f(a)
const N2 = f => a => f(f(a))
const N3 = f => a => f(f(f(a)))
const N4 = f => a => f(f(f(f(a))))
const N5 = f => a => f(f(f(f(f(a)))))

//lists
const nil = pair(T)(T)
const isNil = p => p(K)
const cons = h => t => pair(F)(pair(h)(t))
const head = z => first(second(z))
const tail = z => second(second(z))

//trees
const Empty = e => b => e
const Branch = x => l => r => e => b => b(x)(l(e)(b))(r(e)(b))
const Treesum = t => t(0)(x => r => l => x+l+r)